trinity/sync/full/trie.py

```diff
@@ -0,0 +1,231 @@
+        elif isinstance(node[1], list):
+            # the rlp encoding of the node is < 32 so rather than a 32-byte
+            # reference, the actual rlp encoding of the node is inlined.
+            sub_references, sub_leaves = _get_children(node[1], depth + 1)
+            references.extend(sub_references)
+            leaves.extend(sub_leaves)
+        else:
+            raise Exception("Invariant")
+    elif node_type == NODE_TYPE_BRANCH:
+        for sub_node in node[:16]:
+            if isinstance(sub_node, bytes) and len(sub_node) == 32:
+                # this is a reference to another node.
+                references.append((depth + 1, sub_node))
+            else:
+                sub_references, sub_leaves = _get_children(sub_node, depth)
+                references.extend(sub_references)
+                leaves.extend(sub_leaves)
+
+        # The last item in a branch may contain a value.
+        if not is_blank_node(node[16]):
+            leaves.append(node[16])
+
+    return references, leaves
+
+
+class HexaryTrieSync:
+
+    def __init__(self, root_hash: Hash32, db: AsyncBaseDB, logger: TraceLogger) -> None:
+        self.queue: List[SyncRequest] = []
+        self.requests: Dict[Hash32, SyncRequest] = {}
+        self.db = db
+        self.root_hash = root_hash
+        self.logger = logger
+        # A cache of node hashes we know to exist in our DB, used to avoid querying the DB
+        # unnecessarily as that's the main bottleneck when dealing with a large DB like for
+        # ethereum's mainnet/ropsten.
+        self._existing_nodes: Set[Hash32] = set()
+        self.committed_nodes = 0
+        if root_hash in self.db:
+            self.logger.info("Root node (%s) already exists in DB, nothing to do", root_hash)
+        else:
+            self._schedule(root_hash, parent=None, depth=0, leaf_callback=self.leaf_callback)
+
+    async def leaf_callback(self, data: bytes, parent: SyncRequest) -> None:
+        """Called when we reach a leaf node.
+
+        Should be implemented by subclasses that need to perform special handling of leaves.
+        """
+        pass
+
+    @property
+    def has_pending_requests(self) -> bool:
+        return len(self.requests) > 0
+
+    def next_batch(self, n: int = 1) -> List[SyncRequest]:
+        """Return the next requests that should be dispatched."""
+        if len(self.queue) == 0:
+            return []
+        batch = list(reversed((self.queue[-n:])))
+        self.queue = self.queue[:-n]
+        return batch
+
+    async def schedule(self, node_key: Hash32, parent: SyncRequest, depth: int,
+                       leaf_callback: Callable[[bytes, 'SyncRequest'], Awaitable[None]],
+                       is_raw: bool = False) -> None:
+        """Schedule a request for the node with the given key."""
+        if node_key in self._existing_nodes:
+            self.logger.debug("Node %s already exists in db" % encode_hex(node_key))
```

Comment:
Logging statements should not do string formatting.  Can you change to `self.logger.debug("the %s message", encode_hex(nodekey))`